### PR TITLE
[11.x] Upgrade to Carbon v3

### PR DESCRIPTION
**Pull request by kylekatarnls**
- Update Carbon to `^3@dev`, if needed 1 Laravel version could allow `^2.72.2 || ^3@dev` to transition but would also implies some complexity to support both
- Removed setTestNow override, in Carbon 3 it's synchronized already to the PSR clock
- Noticing createFromId PHPDoc says it accepts Uuid and Ulid but code fails if not string so I changed it to respect its PHPDoc

### CODE REVIEW

**Comment by driesvints** [comment]
Hey @kylekatarnls. When were you planning on releasing v3? It'd be best that's done before Laravel v11 is released.

**Comment by kylekatarnls** [comment]
I planned to check with you if Laravel would need any adjustment in this version so to try to detect as much as possible before releasing to stable.

On my side it's packed and if you don't have any major concern, I will release a beta tag this week.

Here are the main changes for [3.0](https://github.com/briannesbitt/Carbon/milestones/3.0.0)

- Drop PHP < 8.1
  - briannesbitt/Carbon#2810
  - briannesbitt/Carbon#2385
  - briannesbitt/Carbon#2070
  - briannesbitt/Carbon#2346
- briannesbitt/Carbon#2885
- briannesbitt/Carbon#1752
- briannesbitt/Carbon#2340
- briannesbitt/Carbon#2035
- briannesbitt/Carbon#2079
- briannesbitt/Carbon#2119
  - diffIn* will use the floatDiffInReal* behavior, all other variants will be removed
- briannesbitt/Carbon#1967
- briannesbitt/Carbon#2347
- briannesbitt/Carbon#2519
- briannesbitt/Carbon#2701
- briannesbitt/Carbon#2345

**Comment by driesvints** [comment]
@kylekatarnls Right. I'll let @taylorotwell review the code changes here.

Just want to say that we'll need a stable Carbon v3 tag before we release Laravel v11 otherwise we can't get this in. We can't release with a dev stability package.

File: src/Illuminate/Support/Carbon.php
Comment (crynobone) [inline src/Illuminate/Support/Carbon.php:27]: > Noticing createFromId PHPDoc says it accepts Uuid and Ulid but code fails if not string so I changed it to respect its PHPDoc

Doesn't the same code be applied to Laravel 10 and Carbon 2.x?

File: src/Illuminate/Support/Carbon.php
Comment (crynobone) [inline src/Illuminate/Support/Carbon.php:23]:
```suggestion
        if (is_string($id)) {
```

File: composer.json
Comment (crynobone) [inline composer.json:38]: We need to update `src/Illuminate/Support/composer.json` as well.

**Review by crynobone — CHANGES_REQUESTED** [review]

File: src/Illuminate/Support/Carbon.php
Comment (kylekatarnls) [reply src/Illuminate/Support/Carbon.php:27]: Yes I can make a separated PR for that.

File: tests/Integration/Console/Scheduling/SubMinuteSchedulingTest.php
Comment (crynobone) [inline tests/Integration/Console/Scheduling/SubMinuteSchedulingTest.php:140]: Changes to this file can be applied to Carbon 2 and Laravel 10, right?

**Comment by taylorotwell** [comment]
@kylekatarnls can you summarize any significant breaking changes for us?

File: tests/Integration/Console/Scheduling/SubMinuteSchedulingTest.php
Comment (kylekatarnls) [reply tests/Integration/Console/Scheduling/SubMinuteSchedulingTest.php:140]: Yes. 👍

**Comment by kylekatarnls** [comment]
# diffIn*
Yes, the most impactful change is in `diffIn*` methods. They was returning positive integer by default, they will now return float:
```php
$after = Carbon::now()->addSeconds(2);
$before = Carbon::now();

var_dump($after->diffInSeconds($before));
```
This was: `int(1)` in Carbon 2
This is: `double(-1.999508)` in Carbon 3

In Carbon 3, using `(int) $after->diffInSeconds($before, true)` or `(int) abs($after->diffInSeconds($before))` allows to get explicitly an absolute and truncated value, so the same result as in v2.

# Strong typing
Strong typing added in many method parameters will disallow so usages that was never meant to be supported such as `bool` or `null` in comparison methods:
In Carbon v2 the following return arbitrary results (despite it emits `E_USER_DEPRECATED` notices):
```php
var_dump($date->eq(false)); // false
var_dump($date->gte(false)); // true
var_dump($date->gt(false)); // true
var_dump($date->eq(null)); // false
var_dump($date->gte(null)); // true
var_dump($date->gt(null)); // true
```
In Carbon v3, PHPDoc is moving to real types so any of this will throw a `TypeError`, it expects `DateTimeInterface` or `string` to be passed.

# falsable => nullable
`create`, `createFromFormat` , `createFromIsoFormat`, `createFromLocaleFormat` and `createFromLocaleIsoFormat` returned `Carbon|false`, it's now `Carbon|null` (unlocking `?->` chaining and `?? throw` error handling).

# Removal of some deprecated methods
`setUtf8` are `formatLocalized` removed. `formatLocalized` depends on OS language package and relies on `setlocale()` + `strftime()` (the last will be removed from PHP itself). `->isoFormat()` should be used instead which is embedded translation files (so one can rely to get the same translation everywhere) and is linked to Carbon locale (The current object locale that can be set with `$date->locale('de')` or if not set, the global one that Laravel is already setting using `Carbon::setLocale('de')`). And `setUtf8` is no longer needed since `formatLocalized` is the only method that possibly had non-UTF-8 output. This might be a BC for some users, but I doubt that any Laravel package is using that internally.

No other point should be relevant for Laravel.

File: src/Illuminate/Support/Carbon.php
Comment (kylekatarnls) [reply src/Illuminate/Support/Carbon.php:27]: Extracted: https://github.com/laravel/framework/pull/49783

File: tests/Integration/Console/Scheduling/SubMinuteSchedulingTest.php
Comment (crynobone) [reply tests/Integration/Console/Scheduling/SubMinuteSchedulingTest.php:140]: Extracted #49785

**Comment by crynobone** [comment]
@kylekatarnls can you merge `master` from upstream to this PR.

**Review by crynobone — APPROVED** [review]

**Review by driesvints — APPROVED** [review]

**Comment by driesvints** [comment]
@taylorotwell let's merge this one and then we can see if any of the nightly builds of our packages will fail.

**Comment by kylekatarnls** [comment]
I'm looking at the tests, I think it's on Carbon-side and needs a new beta tag

**Comment by driesvints** [comment]
Thanks @kylekatarnls. Converted this to draft until we figured that one out.

**Comment by GrahamCampbell** [comment]
Giving the CI a kick to pick up the new carbon tag.

**Comment by kylekatarnls** [comment]
@driesvints Is it fine to merge with beta version so some people having Laravel on dev-master will start to test Carbon 3 in Laravel context and may raise points not covered by tests? Of course I would release a stable tag before Laravel tags its own stable version 11

**Comment by driesvints** [comment]
@kylekatarnls yep, perfect 👍

**Pull request merged**
